controllers/awscluster_controller.go

```diff
@@ -66,9 +77,41 @@ var (
 // AWSClusterReconciler reconciles a AwsCluster object.
 type AWSClusterReconciler struct {
 	client.Client
-	Recorder         record.EventRecorder
-	Endpoints        []scope.ServiceEndpoint
-	WatchFilterValue string
+	Recorder              record.EventRecorder
+	ec2ServiceFactory     func(scope.EC2Scope) services.EC2MachineInterface
+	networkServiceFactory func(scope.ClusterScope) services.NetworkInterface
+	elbServiceFactory     func(scope.ClusterScope) services.ELBClusterInterface
+	securityGroupFactory  func(scope.ClusterScope) services.SecurityGroupInterface
+	Endpoints             []scope.ServiceEndpoint
+	WatchFilterValue      string
+}
+
+func (r *AWSClusterReconciler) getEC2Service(scope scope.EC2Scope) services.EC2MachineInterface {
+	if r.ec2ServiceFactory != nil {
```

Comment:
It'd be good to have comments for all these factories to make it clear that these are added for enabling testing.

Reply:
done